torch/distributions/one_hot_categorical.py

```diff
@@ -0,0 +1,60 @@
+import torch
+from torch.autograd import Variable
+from torch.distributions.distribution import Distribution
+from torch.distributions.categorical import Categorical
+
+
+class OneHotCategorical(Distribution):
+    r"""
+    Creates a one-hot categorical distribution parameterized by `probs`.
+
+    Samples are one-hot coded vectors of size probs.size(-1).
+
+    See also: :func:`torch.distributions.Categorical`
+
+    Example::
+
+        >>> m = OneHotCategorical(torch.Tensor([ 0.25, 0.25, 0.25, 0.25 ]))
+        >>> m.sample()  # equal probability of 0, 1, 2, 3
+         0
+         0
+         1
+         0
+        [torch.FloatTensor of size 4]
+
+    Args:
+        probs (Tensor or Variable): event probabilities
+    """
+    has_enumerate_support = True
+
+    def __init__(self, probs):
+        self._categorical = Categorical(probs)
+        batch_shape = probs.size()[:-1]
+        event_shape = probs.size()[-1:]
+        super(OneHotCategorical, self).__init__(batch_shape, event_shape)
+
+    def sample(self, sample_shape=torch.Size()):
+        sample_shape = torch.Size(sample_shape)
+        probs = self._categorical.probs
+        one_hot = probs.new(sample_shape + probs.size()).zero_()
```

Comment:
nit: `probs.new(self._extended_shape(sample_shape)).zero_()` for consistency with the other distributions. 

Also, this will make the tensor type be `float` or `double`. Is that something that we want? I think it is fine to have this and let the user cast it using `.long()` if they need to use this for indexing (keeps the return types simple), but I thought we were doing this casting inside the distributions in `Pyro`.

Reply:
Yes, one-hot tensors are usually of float type, whereas the index tensors of `Categorical` are more naturally of long type (since the only thing you can do with them is index into other tensors).